### PR TITLE
[processor/probabilisticsampler] Log sampling

### LINKED ISSUE #9117: Support log sampling
**Is your feature request related to a problem? Please describe.**
Similar to what is offered with Probabilistic Sampling Processor, offer a log sampler.

**Describe the solution you'd like**
Ideally, the exact same config as the trace equivalent, with the addition of being able to sample logs according to their severity.

**Describe alternatives you've considered**
None come to mind.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by atoulme**
**Description:** 
Add a probabilistic log sampler

**Link to tracking Issue:**
Closes #9117

**Testing:**
Unit tests

**Documentation:**
README

### CODE REVIEW

**Comment by atoulme** [comment]
This PR reprises #9118.

File: processor/probabilisticsamplerprocessor/README.md
Comment (jpkrohling) [inline processor/probabilisticsamplerprocessor/README.md:38]: This looks very much like the one for traces. Perhaps this readme should be split into sections, one for the common characteristics, one for traces, and one for logs.

File: processor/probabilisticsamplerprocessor/README.md
Comment (jpkrohling) [inline processor/probabilisticsamplerprocessor/README.md:69]: How about something like:

```
processors:
  probabilistic_sampler:
    sampling_percentage: 15
    attribute_source: record # possible values: one of context, resource, traceID, record
    from_attribute: logID # value is required if the source is not traceID  
```

At first, you could have `record` and `traceID` only, expanding to other sources in the future. This nomenclature is inline with other processors, like the routing processor: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/routingprocessor#configuration

File: processor/probabilisticsamplerprocessor/factory.go
Comment (jpkrohling) [inline processor/probabilisticsamplerprocessor/factory.go:48]: The logs processor should be Alpha, given there are no known production users of this feature at this point.

File: processor/probabilisticsamplerprocessor/logprobabilisticsampler.go
Comment (jpkrohling) [inline processor/probabilisticsamplerprocessor/logprobabilisticsampler.go:26]:
```suggestion
type logSamplerProcessor struct {
```

File: processor/probabilisticsamplerprocessor/logprobabilisticsampler.go
Comment (jpkrohling) [inline processor/probabilisticsamplerprocessor/logprobabilisticsampler.go:78]: Can you move the hash function to another source file? And perhaps rename the current `probabilisticsampler.go` to `tracesprocessor.go`?

File: processor/probabilisticsamplerprocessor/logprobabilisticsampler.go
Comment (jpkrohling) [inline processor/probabilisticsamplerprocessor/logprobabilisticsampler.go:1]: Perhaps this file should be named `logsprocessor.go`.

File: processor/probabilisticsamplerprocessor/logprobabilisticsampler.go
Comment (jpkrohling) [inline processor/probabilisticsamplerprocessor/logprobabilisticsampler.go:79]: The traces processor records metrics with the decision. I think the logs processor should do the same.

File: processor/probabilisticsamplerprocessor/README.md
Comment (atoulme) [reply processor/probabilisticsamplerprocessor/README.md:69]: This is nicer. Adopting it.

File: processor/probabilisticsamplerprocessor/logprobabilisticsampler.go
Comment (atoulme) [reply processor/probabilisticsamplerprocessor/logprobabilisticsampler.go]: added something to that effect.

File: processor/probabilisticsamplerprocessor/config.go
Comment (jpkrohling) [inline processor/probabilisticsamplerprocessor/config.go:48]: If you are changing this PR for any other reason, please add an empty line between lines 47 and 48, and between 50-51.

File: processor/probabilisticsamplerprocessor/logsprocessor.go
Comment (jpkrohling) [inline processor/probabilisticsamplerprocessor/logsprocessor.go:64]: This logic will always yield the same result for the entire processor lifecycle, won't it? It might be better to run this during the startup phase, storing the result instead of recalculating it here, which is the hot path.

File: processor/probabilisticsamplerprocessor/murmur3_test.go
Comment (jpkrohling) [inline processor/probabilisticsamplerprocessor/murmur3_test.go:28]: Is this just ensuring there are no collisions? If so, the test name could be "TestHashHasNoCollisions" (which is impossible to prove).

**Review by jpkrohling — COMMENTED** [review]
LGTM. Factoring out the attribute source resolution from the hot path is worth a change. That being done, this is ready to be merged.

File: processor/probabilisticsamplerprocessor/logsprocessor.go
Comment (atoulme) [reply processor/probabilisticsamplerprocessor/logsprocessor.go:64]: no, this logic depends on the log record attributes.
First, we check on `l.TraceID()`, and then on `l.Attributes().Get(lsp.samplingSource)`, and finally `l.Attributes().Get(lsp.samplingPriority)`. While we have a configuration that is global, we can configure the sampler to pick the field to use for sampling decision and the priority dynamically.

File: processor/probabilisticsamplerprocessor/murmur3_test.go
Comment (atoulme) [reply processor/probabilisticsamplerprocessor/murmur3_test.go:28]: This is just a refactor from the tests that were there before as we moved the hashing algo to its own file. I can certainly rename it.

File: processor/probabilisticsamplerprocessor/README.md
Comment (mx-psi) [reply processor/probabilisticsamplerprocessor/README.md:5]: The OTLP receiver has a separate row per signal https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/README.md

Comment (atoulme) [reply processor/probabilisticsamplerprocessor/README.md]: Changed to adopt that format.

**Comment by atoulme** [comment]
@jpkrohling would this look good to approve for you?

File: processor/probabilisticsamplerprocessor/logsprocessor.go
Comment (harshit-splunk) [inline processor/probabilisticsamplerprocessor/logsprocessor.go:79]: It should ignore `localPriority` when it is not a double value. Otherwise, it will always have 0 priority. WDYT?

Comment (atoulme) [reply processor/probabilisticsamplerprocessor/logsprocessor.go:79]: Makes sense. I think we could land this and you should take it from there.

File: processor/probabilisticsamplerprocessor/murmur3.go
Comment (MovieStoreGuy) [inline processor/probabilisticsamplerprocessor/murmur3.go:18]: Would it be worth important a maintained library here so we are not reimplementing known code?

For example: https://github.com/twmb/murmur3

Comment (atoulme) [reply processor/probabilisticsamplerprocessor/murmur3.go:18]: Not sure - here I'm just moving code around, and I don't want to break existing functionality for traces. It might make sense to follow up on this separately.

File: processor/probabilisticsamplerprocessor/logsprocessor_test.go
Comment (MovieStoreGuy) [inline processor/probabilisticsamplerprocessor/logsprocessor_test.go:70]: Could you please also add in a check for `Nil` here for `got`?

**Review by MovieStoreGuy — CHANGES_REQUESTED** [review]
I have left some comments that I'd like to hear your thoughts on. 

I believe there is one point in which this processor can panic which should be addressed, and I have hesitations on using `Attributes` as a means to power logic within a generic processor.

**Comment by harshit-splunk** [comment]
I'll take over the PR from now.

File: processor/probabilisticsamplerprocessor/config.go
Comment (harshit-splunk) [reply processor/probabilisticsamplerprocessor/config.go:53]: Done

File: processor/probabilisticsamplerprocessor/config.go
Comment (harshit-splunk) [reply processor/probabilisticsamplerprocessor/config.go:55]: Traces will not be affected as these configs are used in the log processor only.

File: processor/probabilisticsamplerprocessor/logsprocessor.go
Comment (harshit-splunk) [reply processor/probabilisticsamplerprocessor/logsprocessor.go:81]: Spans can be sampled based on the `sampling.priority` attribute. We intend to do similar for logs data.
I think it can be a good starting point for adding logs sampling.

File: processor/probabilisticsamplerprocessor/logsprocessor_test.go
Comment (harshit-splunk) [reply processor/probabilisticsamplerprocessor/logsprocessor_test.go:71]: Done

File: processor/probabilisticsamplerprocessor/README.md
Comment (jpkrohling) [reply processor/probabilisticsamplerprocessor/README.md:5]: Yes, let's be explicit on what each signal is capable of and with which maturity level. Just make sure to follow the same formatting.

File: processor/probabilisticsamplerprocessor/logsprocessor_test.go
Comment (jpkrohling) [inline processor/probabilisticsamplerprocessor/logsprocessor_test.go:86]: Will that commented-out code eventually be uncommented? It might be better to remove them now and add them later, once they are needed.

File: processor/probabilisticsamplerprocessor/murmur3.go
Comment (jpkrohling) [reply processor/probabilisticsamplerprocessor/murmur3.go:18]: I created https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16456 to track this.

**Review by jpkrohling — COMMENTED** [review]
I would love to see the commented-out gone before merging this PR. Otherwise, LGTM.

**Review by jpkrohling — APPROVED** [review]
@MovieStoreGuy, would you mind taking a final look? I can't merge without your approval.

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by MovieStoreGuy** [comment]
Sorry for the delay on getting back here.

Once the conflicts have been resolved, it should be easy to get through :)

**Pull request merged**
